NAP-6: Updates after discussion and review (#312)
napari/docs: docs/naps/6-contributable-menus.md

one of these menu items.

1. `Visualize` - Items in this submenu allow you to generate visualizations from selected layer or layers.
They do not change the layer data.
2. `Measure` - Items in this submenu provide utilities for summarising information about your layer's data.
3. `Edit` - The items in this submenu change the data of your layer through `Filters` or `Transformations`. 
Additionally `Annotation Tools` provide a location for convenience layer editing tools e.g. `Labels` split/merge actions.
Items in this submenu **should not** generate new layers, but rather act upon the existing layer data.

The remaining submenus under `Layers` deal with generating new layers based on processing of existing layers
or other information. They define common *analysis* actions users may want to take on their layers.

4. `Registration` - Commands that allow the user to perform image registration on one or more layers
5. `Projection` - Commands that generate various projections based on one or more layers
6. `Segmentation` - Commands that generate image segmentations
7. `Tracks` - Commands that take in an image or segmentation and produce timelapse tracks

> 8. `Classification` - ??

Reviewer: Why the "??"?

2. `Measure` - Items in this submenu provide utilities for summarising information about your layer's data. An example
would be a widget that plots the change in the intensity of a clicked pixel over time.
3. `Edit` - The items in this submenu allow you to change the data of your layer through `Filter` actions e.g. a gaussian 
filter or `Transform` actions e.g. an affine transform. Additionally `Annotate` provides a location for convenient layer 
editing tools e.g. `Labels` split/merge actions. Items in this submenu **should not** generate new layers, but rather act upon 
the existing layer data.

The remaining submenus under `Layers` deal with generating new layers based on processing of existing layers
or other information. They define common *analysis* actions users may want to take on their layers.

4. `Register` - Commands that allow the user to perform image registration on one or more layers
5. `Project` - Commands that generate various projections based on one or more layers
6. `Segment` - Commands that generate instance segmentations e.g. watershed, SLIC, threshold
7. `Track` - Commands that take in an image or segmentation and produce timelapse tracks

> 8. `Classify` - Commands that predict classes/categories for pixels, objects, points, etc.

Reviewer: Question, would these be submenus in `Layers` or would they one layer deeper e.g., `Layers/Processing/...` ? (just thinking about how long the `Layers` menu would be)

Author: These are in `Layers` itself, after a separator. I did initially have them under `Layers/Generate` (with the idea being that they "generate" new layers from existing layers, but folks preferred to just move everything out to the top level)

> Many of the actions in this menu exist in the right click layer context menu. These items

Reviewer: Question - would all items in the layer context menu be replicated in the `Layers` menubar menu?

Author: Great question! I imagine they will all somehow be replicated either somewhere in the `Layers` menu or somewhere in the `Edit` menu? But again not super clear to me just now what the distinction between those should be, so I want to have a closer look at e.g. Photoshop

should be replicated in the `Layers` menu as needed, both to aid discoverability and
to ensure users are not met with empty menus on initially opening `napari`. It's
also possible that exceedingly common operations e.g. thresholds, are provided in the future
by `napari` itself.

### The `Acquire` Menu

> In addition to the `Layers` menu, we add `Acquire` as a top level menu.

Reviewer: ```suggestion
In addition to the `Layers` menu, we would add `Acquire` as a top level menu.
```

exporting various file formats. These interfaces will be exposed via the new `File->I/O Utilities`
menu.

### The `File -> New Layer` Menu

This menu will contain built-in `napari` commands for creating new empty layers (similar to 

> the new layer buttons atop the layerlist), but also allow plugins to declare commands that 

Reviewer: ```suggestion
the 'new layer' buttons atop the layerlist), but also allow plugins to declare commands that 
```

data formats or layer types.

We therefore give plugin developers full control over their own submenu under `Plugins->My Plugin`.
Plugin developers can organize all contributions under this submenu as they see fit, including adding 
their own submenus of arbitrary depth.

All commands offered by a plugin will always be available in the plugin's submenu, regardless of
their presence in other menus. If the plugin defines its own submenus, these will always be first,
followed by a separator, then any commands with no dedicated menu entry for the plugin's own 

> submenu. These will be alphabetically ordered.

Reviewer: I think I forgot this detail. Are we sure we want to duplicate menu items (i.e., place contribution in the plugin's submenu *and* in another menu, defined by the plugin?)?

Reviewer 2: No, we are not sure, but personally I think it is better to have a way to see all of a plugin's contributions in one place. But we are definitely not sure, this is 100% open to discussion.

```
File
├─ ...
├─ New Layer
├─ IO Utilities
Layers
├─ Visualize
├─ Measure
├─ Edit
│  ├─ Annotate
│  ├─ Filter
│  ├─ Transform
├─ Registration
├─ Projection
├─ Segmentation
├─ Tracks           # remove?
├─ Classification   # remove?
Acquisition
```

> As a case study, we take four plugins offering between 9 and 14 widget contributions and arrange their widgets in these menus: 

Reviewer: nitpick, is that a trailing white space?
```suggestion
As a case study, we take four plugins offering between 9 and 14 widget contributions and arrange their widgets in these menus: 
```

Reviewer: I think there are a few added in this PR.

Author: ok I think I've removed them all!

├─ Measure
│  ├─ SParticlesProperties (napari-stracking)
│  ├─ STracksFeatures (napari-stracking)
├─ Registration
├─ Projection
├─ Segmentation
│  ├─ 2D Inference (empanada-napari)
│  ├─ 3D Inference (empanada-napari)
│  ├─ make_log_segmentation (napari-clemreg)
│  ├─ make_clean_binary_segmentation (napari-clemreg)
│  ├─ SegmentOtsu (napari-mm3)
│  ├─ SegmentUnet (napari-mm3)
│  ├─ Foci (napari-mm3)
│  ├─ SDetectorDog (napari-stracking)
│  ├─ SDetectorDoh (napari-stracking)
│  ├─ SDetectorLog (napari-stracking)
│  ├─ SDetectorSeg (napari-stracking)
├─ Tracks
│  ├─ SLinkerShortestPath (napari-stracking)
│  ├─ Tracks (napari-mm3)
├─ Classification
├─ make_point_cloud_sampling (napari-clemreg)

> ├─ make_point_cloud_registration (napari-clemreg)

Reviewer 2: why is this not in Registration?

Author: Iunno... it is now 🙈

│  ├─ predict_from_model
```

### Item Descriptions

To allow users to browse menus more effectively without having to click each
item to figure out what it does, menu contributions should be given a `description`
field that is available when the user is inspecting the menu. The `description`

> should be required, so that 

Reviewer: To be finished.
Also just to clarify we would want description to be 'tooltip' field in `Action` which appears in the bottom status bar right?

Author: Done.

### Item Grouping & Ordering

Previously, plugin contributions were limited to the `Plugins` menu and grouped
under the plugin's name. Now that plugin contributions can be colocated with 
native napari actions, it's important that users are able to distinguish
the source of menu items.

To that end, napari items should always be grouped separately to plugin items
in all menus. Additionally, the plugin's name should also be listed with 
each plugin contribution. Since plugin names can be quite long, future work should

> consider more concise ways to indicate menu item sources, including using icons.

Reviewer: If we do have an icon to indicate plugin contribution, we could add the plugin name to 'description' (i.e. tooltip field, so it appears on hover)

Reviewer 2: I like this idea. Potential icon: 🔌

Author: Added a sentence.

Outside of a plugin's own submenu, the `order` and `group` keys will be ignored

> for menu contributions. napari will make its own decisions about the grouping

Author: what is napari strat

Author: Expanded.

The user proposing the menu is not responsible for opening this Pull
Request, though they may, if they wish.

### User Configuration

This document has so far discussed menu contributions exclusively from
the plugin developer's perspective. In general, napari will continue
to be opinionated about what plugin developers can and cannot change 
about the napari interface through menu contributions. Our guiding principle
here is that users should never be surprised by the napari interface as a result
of standard plugin contributions - menus shouldn't move around, or be hidden, layer
controls should behave in the same way regardless of what plugins are installed, etc.

However, napari users **should** be able to freely customize their own napari installation
to a much greater extent. This should include contribution-level enablement control

> for plugins, as well as building their own custom menus, whether this be a brand new top-level

Reviewer: Naive question, how would users control plugin contribution enablement..?

Reviewer 2: I imagine, a tree selection widget like this one:

![flutter tree view](https://user-images.githubusercontent.com/22675676/136653434-133b7e4a-fa57-463c-a11a-64a8711c9de9.gif)

I think there should be some internal npe2 data structure for this, but the UI itself should itself be a plugin. (Maybe built-ins.)

Reviewer: Would this be separate to our plugin manager (i.e. the install/uninstall/enable/disable dialog)?

(plugin manager has several meanings in napari: https://github.com/napari/docs/issues/321#issuecomment-2011580837 :see_no_evil: )

Reviewer 2: I'm ambivalent about it, and would leave that decision to whoever ends up implementing it.

On the one hand, it can be an independent dialog — enablement of contributions only needs npe2 and stuff already installed, and doesn't need to depend on conda, pip, etc. So you eliminate a lot of complexity. Additionally, I personally would never use the plugin manager (I directly install things with micromamba as needed), but I would probably use the enablement UI because the alternative, I think, is editing some random yaml or something? 😂 Which I'm not super excited about.

On the other hand, I can imagine a user being in the enablement UI and thinking hmm, I really don't use this plugin, I should remove it. And they would probably be grateful if that can be done in the same widget rather than have to launch a totally different interface.

## Future Work

> **Command paltte:** As mentioned above, a key feature to support rapid browsing for actions is the

Reviewer: ```suggestion
**Command palette:** As mentioned above, a key feature to support rapid browsing for actions is the
```

## Future Work

**Command palette:** As mentioned above, a key feature to support rapid browsing for actions is the
search functionality via the command palette. This is actively being worked
on and is essential for navigation.

**Context dependent enablement:** Once more contribution types are exposed for users, it's important that users
are aware why certain actions are disabled when the user doesn't meet the
requisite context declared in the contribution's `enablement` clause. Since
the syntax for declaring these contexts is strictly defined, we should be
able to surface information to users about what is required for the action
to be enabled and functional. For example, an action could declare itself
enabled only when a points layer *and* an image layer are selected. If 
the user has only selected an image layer, we could indicate the missing
context to the user e.g. "Action takes a points and image layer, but no

> points layer is selected". 

Reviewer: ```suggestion
points layer is selected".
```

nit - trailing whitespace. Noted below too..
